hal/ia32/cpu.c

```diff
@@ -135,62 +135,37 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
 {
 	int err;
 
-	if (spinlock != NULL) {
-		hal_cpuGetCycles((void *)&spinlock->e);
-
-		/* Calculate maximum and minimum lock time */
-		if ((cycles_t)(spinlock->e - spinlock->b) > spinlock->dmax)
-			spinlock->dmax = spinlock->e - spinlock->b;
-
-		if (spinlock->e - spinlock->b < spinlock->dmin)
-			spinlock->dmin = spinlock->e - spinlock->b;
-	}
-
-	__asm__ volatile (
-		"movl %1, %%eax;"
-		"cmp $0, %%eax;"
-		"je 1f;"
-
-		"movl %3, %%eax;"
-		"pushl %%eax;"
-		"xorl %%eax, %%eax;"
-		"incl %%eax;"
-		"xchgl %2, %%eax;"
-		"jmp 2f;"
-
-		"1:;"
-		"pushf;"
-
-		"2:;"
-		"pushl %%cs;"
-		"cli;"
-		"leal 3f, %%eax;"
-		"pushl %%eax;"
-		"movl $0, %%eax;"
-		"call interrupts_pushContext;"
-
-		"call _interrupts_multilockSet;"
-
-		"leal 0(%%esp), %%eax;"
-		"pushl $0;"
-		"pushl %%eax;"
-		"pushl $0;"
-		"movl %4, %%eax;"
-		"call *%%eax;"
-		"cli;"
-		"addl $12, %%esp;"
-		"popl %%esp;"
-
-		"call _interrupts_multilockClear;"
- 		"pushl %%esp;"
-
-		"jmp interrupts_popContext;"
-
-		"3:;"
-		"movl %%eax, %0"
-	: "=g" (err)
-	: "m" (spinlock), "m" (spinlock->lock), "r" (*scp), "g" (threads_schedule)
-	: "eax", "edx", "esp", "cc", "memory");
+	hal_cpuDisableInterrupts();
```

Comment:
This won't work - spinlockClear will enable interrupts by restoring interrupt mask. That's why we need to do spinlockClear manually in reschedule - so we'll only clear lock without restoring interrupt mask

Reply:
@agkaminski I see your point - `popf` in `hal_spinlockClear()` enables interrupts back, so I've restored manual spinlock clearing. Please take a look at the code again.

Sidenote: the original problem with `hal_cpuReschedule()` was passing "r" (*scp) in the input list. It caused page fault (NULL dereference) when calling `hal_cpuReschedule(NULL, NULL)`. The `*scp` dereference (red rectangle) got executed before `spinlock != NULL` check (`cmp $0x0, %eax` below the rectangle).
![reschedule](https://user-images.githubusercontent.com/8835317/135496578-1b076835-4d77-4246-81d5-bc418ae5710f.png)